libbpf-tools/profile.c

```diff
@@ -0,0 +1,649 @@
+					 pid_t pid)
+{
+	int i;
+	const struct sym *sym;
+	const struct syms *syms = syms_cache__get_syms(syms_cache, pid);
+
+	if (!syms)
+		return;
+
+	for (i = env.perf_max_stack_depth - 1; i >= 0; i--) {
+		if (ip[i] == 0)
+			continue;
+
+		sym = syms__map_addr(syms, ip[i]);
+		printf(";%s", sym ? sym->name : "[unknown]");
+	}
+}
+
+static void print_kernel_stacktrace(unsigned long *ip, struct ksyms *ksyms)
+{
+	int i;
+	const struct ksym *ksym;
+
+	for (i = 0; ip[i] && i < env.perf_max_stack_depth; i++) {
+		ksym = ksyms__map_addr(ksyms, ip[i]);
+		printf("    %s\n", ksym ? ksym->name : "unknown");
+	}
+}
+
+static void print_kernel_stacktrace_folded(unsigned long *ip, struct ksyms *ksyms)
+{
+	int i;
+	const struct ksym *ksym;
+
+	for (i = env.perf_max_stack_depth - 1; i >= 0; i--) {
+		if (ip[i] == 0)
+			continue;
+
+		ksym = ksyms__map_addr(ksyms, ip[i]);
+		printf(";%s", ksym ? ksym->name : "[unknown]");
+	}
+}
+
+static void print_count(struct key_t *event, __u64 count, int sfd,
+			struct ksyms *ksyms, struct syms_cache *syms_cache)
+{
+	unsigned long *ip;
+
+	ip = calloc(env.perf_max_stack_depth, sizeof(*ip));
+	if (!ip) {
+		fprintf(stderr, "failed to alloc ip\n");
+		return;
+	}
+
+	/* kernel stack */
+	if (!env.user_stacks_only && !STACK_ID_EFAULT(event->kern_stack_id)) {
+		if (bpf_map_lookup_elem(sfd, &event->kern_stack_id, ip) != 0)
+			printf("    [Missed Kernel Stack]\n");
+		else
+			print_kernel_stacktrace(ip, ksyms);
+	}
+
+	/* user stack */
+	if (!env.kernel_stacks_only && !STACK_ID_EFAULT(event->user_stack_id)) {
+		if (NEED_DELIMITER(env.delimiter, event->user_stack_id, event->kern_stack_id))
+			printf("    --\n");
+
+		if (bpf_map_lookup_elem(sfd, &event->user_stack_id, ip) != 0)
+			printf("    [Missed User Stack]\n");
+		else
+			print_user_stacktrace(ip, syms_cache, event->pid);
+	}
+
+	printf("    %-16s %s (%d)\n", "-", event->name, event->pid);
+	printf("        %lld\n\n", count);
+
+	free(ip);
+}
+
+static void print_count_folded(struct key_t *event, __u64 count, int sfd,
+			       struct ksyms *ksyms, struct syms_cache *syms_cache)
+{
+	unsigned long *ip;
+
+	ip = calloc(env.perf_max_stack_depth, sizeof(*ip));
+	if (!ip) {
+		fprintf(stderr, "failed to alloc ip\n");
+		return;
+	}
+
+	printf("%s", event->name);
+
+	/* user stack */
+	if (!env.kernel_stacks_only && !STACK_ID_EFAULT(event->user_stack_id)) {
+		if (bpf_map_lookup_elem(sfd, &event->user_stack_id, ip) != 0)
+			printf(";[Missed User Stack]");
+		else
+			print_user_stacktrace_folded(ip, syms_cache, event->pid);
+	}
+
+	/* kernel stack */
+	if (!env.user_stacks_only && !STACK_ID_EFAULT(event->kern_stack_id)) {
+		if (NEED_DELIMITER(env.delimiter, event->user_stack_id, event->kern_stack_id))
+			printf(";-");
+
+		if (bpf_map_lookup_elem(sfd, &event->kern_stack_id, ip) != 0)
+			printf(";[Missed Kernel Stack]");
+		else
+			print_kernel_stacktrace_folded(ip, ksyms);
+	}
+
+	printf(" %lld\n", count);
+
+	free(ip);
+}
```

Comment:
> Maybe, we can check the `folded` argument in the `print_*_stacktrace()` functions?
It can reduce duplication slightly, but it cannot reduce meaningfully because iterating direction and format are different. 

In principle, tt's good to separate functions to handle data and representation. For separating this, I can try using iterator that takes addresses and syms and then return resolved symbols, or I can try returning resolved symbols through another array.
But It can make complex code in C language or need more memory usage.

Any thoughts on using iterators for this? Shall I try it?